syntax = "proto3";

package sonet.moderation;

import "google/protobuf/timestamp.proto";
import "google/protobuf/empty.proto";

// Moderation service for Sonet founder system
service ModerationService {
  // Account moderation actions
  rpc FlagAccount(FlagAccountRequest) returns (FlagAccountResponse);
  rpc RemoveFlag(RemoveFlagRequest) returns (RemoveFlagResponse);
  rpc ShadowbanAccount(ShadowbanAccountRequest) returns (ShadowbanAccountResponse);
  rpc SuspendAccount(SuspendAccountRequest) returns (SuspendAccountResponse);
  rpc BanAccount(BanAccountRequest) returns (BanAccountResponse);
  
  // Note moderation actions
  rpc DeleteNote(DeleteNoteRequest) returns (DeleteNoteResponse);
  
  // Moderation queries
  rpc GetFlaggedAccounts(GetFlaggedAccountsRequest) returns (GetFlaggedAccountsResponse);
  rpc GetModerationQueue(GetModerationQueueRequest) returns (GetModerationQueueResponse);
  rpc GetModerationStats(GetModerationStatsRequest) returns (GetModerationStatsResponse);
  
  // Utility methods
  rpc IsFounder(IsFounderRequest) returns (IsFounderResponse);
  rpc CanPerformAction(CanPerformActionRequest) returns (CanPerformActionResponse);
}

// Account flagging
message FlagAccountRequest {
  string target_user_id = 1;
  string target_username = 2;
  string reason = 3;
  string warning_message = 4;
  string founder_id = 5;  // Hidden from users
  string founder_username = 6;  // Hidden from users
}

message FlagAccountResponse {
  bool success = 1;
  string flag_id = 2;
  string message = 3;  // Public message shown to users
  google.protobuf.Timestamp expires_at = 4;
  string error = 5;
}

message RemoveFlagRequest {
  string flag_id = 1;
  string founder_id = 2;
}

message RemoveFlagResponse {
  bool success = 1;
  string message = 2;
  string error = 3;
}

// Account shadowbanning
message ShadowbanAccountRequest {
  string target_user_id = 1;
  string target_username = 2;
  string reason = 3;
  string founder_id = 4;
}

message ShadowbanAccountResponse {
  bool success = 1;
  string message = 2;
  string error = 3;
}

// Account suspension
message SuspendAccountRequest {
  string target_user_id = 1;
  string target_username = 2;
  string reason = 3;
  int32 duration_days = 4;
  string founder_id = 5;
}

message SuspendAccountResponse {
  bool success = 1;
  string message = 2;
  google.protobuf.Timestamp suspension_until = 3;
  string error = 4;
}

// Account banning
message BanAccountRequest {
  string target_user_id = 1;
  string target_username = 2;
  string reason = 3;
  bool permanent = 4;
  string founder_id = 5;
}

message BanAccountResponse {
  bool success = 1;
  string message = 2;
  string error = 3;
}

// Note deletion
message DeleteNoteRequest {
  string note_id = 1;
  string target_user_id = 2;
  string reason = 3;
  string founder_id = 4;
}

message DeleteNoteResponse {
  bool success = 1;
  string message = 2;
  string error = 3;
}

// Query requests
message GetFlaggedAccountsRequest {
  string founder_id = 1;
  bool include_expired = 2;
  int32 limit = 3;
  int32 offset = 4;
}

message GetFlaggedAccountsResponse {
  repeated FlaggedAccount accounts = 1;
  int32 total_count = 2;
  string error = 3;
}

message GetModerationQueueRequest {
  string founder_id = 1;
  string content_type = 2;  // "note", "profile", "user"
  int32 limit = 3;
  int32 offset = 4;
}

message GetModerationQueueResponse {
  repeated ModerationQueueItem items = 1;
  int32 total_count = 2;
  string error = 3;
}

message GetModerationStatsRequest {
  string founder_id = 1;
  google.protobuf.Timestamp period_start = 2;
  google.protobuf.Timestamp period_end = 3;
}

message GetModerationStatsResponse {
  ModerationStats stats = 1;
  string error = 2;
}

// Utility requests
message IsFounderRequest {
  string user_id = 1;
}

message IsFounderResponse {
  bool is_founder = 1;
  string error = 2;
}

message CanPerformActionRequest {
  string user_id = 1;
  string action_type = 2;
}

message CanPerformActionResponse {
  bool can_perform = 1;
  string error = 2;
}

// Data models
message FlaggedAccount {
  string id = 1;
  string user_id = 2;
  string username = 3;
  string reason = 4;
  string warning_message = 5;
  google.protobuf.Timestamp flagged_at = 6;
  google.protobuf.Timestamp expires_at = 7;
  bool is_active = 8;
}

message ModerationQueueItem {
  string id = 1;
  string user_id = 2;
  string username = 3;
  string content_type = 4;
  string content_id = 5;
  string content_preview = 6;
  string flag_reason = 7;
  google.protobuf.Timestamp reported_at = 8;
  google.protobuf.Timestamp expires_at = 9;
  bool is_reviewed = 10;
  bool is_auto_expired = 11;
}

message ModerationStats {
  string id = 1;
  google.protobuf.Timestamp period_start = 2;
  google.protobuf.Timestamp period_end = 3;
  int32 total_flags = 4;
  int32 total_warnings = 5;
  int32 total_shadowbans = 6;
  int32 total_suspensions = 7;
  int32 total_bans = 8;
  int32 total_notes_deleted = 9;
  int32 auto_expired_flags = 10;
  int32 manual_reviews = 11;
}